Document Ephemeral Containers for 1.16 (#15994)
kubernetes/website: content/en/docs/concepts/workloads/pods/ephemeral-containers.md

to build applications.

{{< warning >}}
Ephemeral containers are in early alpha state and are not suitable for production
clusters. You should expect the feature not to work in some situations, such as
when targeting the namespaces of a container. In accordance with the [Kubernetes
Deprecation Policy](/docs/reference/using-api/deprecation-policy/), this alpha
feature could change significantly in the future or be removed entirely.
{{< /warning >}}

{{% /capture %}}

{{% capture body %}}

## Understanding ephemeral containers

{{< glossary_tooltip text="Pods" term_id="pod" >}} are the fundamental building
block of Kubernetes applications. Since Pods are intended to be disposable and
replaceable, you cannot add a container to a Pod once it has been created.
Instead, you usually delete and replace Pods in a controlled fashion using
{{< glossary_tooltip text="deployments" term_id="deployment" >}}.

Sometimes it's necessary to inspect the state of an existing Pod, however, for
example to troubleshoot a hard-to-reproduce bug. In these cases you can run
an ephemeral container in an existing Pod to inspect its state and run
arbitrary commands.

> ### What is an Ephemeral Container?

Reviewer: ```suggestion
### What is an ephemeral container?
```

---
reviewers:
- verb
- yujuhong
title: Ephemeral Containers
content_template: templates/concept
weight: 80
---

{{% capture overview %}}

{{< feature-state state="alpha" >}}

This page provides an overview of ephemeral containers: a special type of container

> that runs temporarily in an existing pod to accomplish user-initiated actions such

Reviewer: ```suggestion
that runs temporarily in an existing {{< glossary_tooltip term_id="pod" >}} to accomplish user-initiated actions such
```

as troubleshooting. You use ephemeral containers to inspect services rather than

> build applications.

Reviewer: ```suggestion
to build applications.
```
(I think this will help localizers and readers for whom English is not their first language)

Author: SG, thanks!

{{< warning >}}
Ephemeral containers are in early alpha state and are not suitable for production

> clusters. You should expect the feature not to work in some situations, like

Reviewer: ```suggestion
clusters. You should expect the feature not to work in some situations, such as
```

when targeting the namespaces of a container. In accordance with the [Kubernetes
Deprecation Policy](/docs/reference/using-api/deprecation-policy/), this alpha
feature could change significantly in the future or be removed entirely.
{{< /warning >}}

{{% /capture %}}

{{% capture body %}}

## Understanding ephemeral containers

{{< glossary_tooltip text="Pods" term_id="pod" >}} are the fundamental building
block of Kubernetes applications. Since pods are intended to be disposable and
replaceable, you cannot add a container to a pod once it has been created.
Instead, you usually delete and replace pods in a controlled fashion using
[deployments](/docs/concepts/workloads/controllers/deployment/).

Sometimes it's necessary to inspect the state of an existing pod, however, for
example to troubleshoot a hard-to-reproduce bug. In these cases you can run
an ephemeral container in an existing pod to inspect its state and run
arbitrary commands.

### Differences from regular containers

> 

Reviewer 2: hello @verb. Thanks for the PR. I went through the files. There are a few nits as well as some general questions.
I wonder if the Ephemeral Containers and Init Containers pages belong under the Containers section.
nit: reword: Differences from regular containers. Possibly: What is an Ephemeral Container
You could then explain that ECs are containers ...

Ephemeral containers lack guarantees for resources or execution, and they will
never be automatically restarted, so they are not appropriate for building
applications.  Ephemeral containers are described using the same `ContainerSpec`
as regular containers, but many fields are incompatible and disallowed for
ephemeral containers.

- Ephemeral containers may not have ports, so fields such as `ports`,
  `livenessProbe`, `readinessProbe` are disallowed.
- Pod resource allocations are immutable, so setting `resources` is disallowed.
- For a complete list of allowed fields, see the [EphemeralContainer reference
  documentation](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#ephemeralcontainer-v1-core).

> 

Reviewer 2: nit: does this link to the reference docs exist? Ref docs for v1.16 have not been generated yet.

Ephemeral containers are created using a special `ephemeralcontainers` handler
in the API rather than by adding them directly to `pod.spec`, so it's not
possible to add an ephemeral container using `kubectl edit`.

> 

Reviewer 2: nit: What is the special handler?

Like regular containers, ephemeral containers may not be changed or removed

> once they are added to the pod.

Reviewer: nit:
```suggestion
once they are added to a Pod.
```

Also consider rewriting this in the active voice.

Author: Updated to active voice. I'm actually a little confused about when I should be capitalizing Pod. Is it ok to use "pod" when referring to the general concept and "Pod" when referring to a specific API object? Likewise with "ephemeral container", should I replace all of these with "EphemeralContainer"?

Reviewer: If you can `kubectl get EphemeralContainer/foo` then write EphemeralContainer in title case.

(You can't, so “ephemeral container” is fine).

For pod vs. Pod, I recommend being consistent. Either capitalize all the uses, or capitalize just the very first use plus anything that's specifically talking about the API.

## Uses for ephemeral containers

Ephemeral containers are useful for interactive troubleshooting when `kubectl
exec` is insufficient because a container has crashed or a container image
doesn't include debugging utilities.

In particular, [distroless images](https://github.com/GoogleContainerTools/distroless)
enable you to deploy minimal container images that reduce attack surface

> and exposure to [CVEs](https://cve.mitre.org/). Since they do not include a

Reviewer: ```suggestion
and exposure to bugs and vulnerabilities. Since distroless images do not include a
```

shell or any debugging utilities, it's difficult to troubleshoot distroless
images using `kubectl exec` alone.

When using ephemeral containers, it's helpful to enable [process namespace
sharing](/docs/tasks/configure-pod-container/share-process-namespace/) so
you can view processes in other containers.

### Examples

> {{< caution >}}

Reviewer: I would use `{{< note >}}` shortcode here as there's no risk to warn readers about.

Author: I started with note, but then noticed that the style guide suggested note was for tips and caution was to avoid pitfalls. This seemed like a pitfall since nothing will work if the feature is disabled.

I don't feel strongly, so happy to do whatever reviewers suggest.

Reviewer: Go for `{{< note >}}`; you can't break your cluster by getting this wrong (I believe?)

> The examples in this section require the `EphemeralContainers` [Feature

Reviewer: nit: “feature gate” should be lower case

Author: fixed

Gate](/docs/reference/command-line-tools-reference/feature-gates.md) to be

> enabled and kubernetes client and server version of at least 1.16.

Reviewer 2: nit: fix link to feature-gates page
nit: fix text/clean-up as mentioned in the task page

Reviewer 2: OK. I understand.

{{< /caution >}}

The examples in this section demonstrate how ephemeral containers appear in
the API. For a more practical example of troubleshooting, see the [Debug a
Running Pod](/docs/tasks/debug-application-cluster/debug-running-pod.md) task.

Ephemeral containers are created using the `ephemeralcontainers` subresource
of pod, which can be demonstrated using `kubectl --raw`. First describe

> the ephemeral container to add as an `EphemeralContainers` list:

Reviewer 2: nit: Pod is a resource

```json
{
    "apiVersion": "v1",
    "kind": "EphemeralContainers",
    "metadata": {
            "name": "example-pod"
    },
    "ephemeralContainers": [{
        "command": [
            "sh"
        ],
        "image": "busybox",
        "imagePullPolicy": "IfNotPresent",
        "name": "debugger",
        "stdin": true,
        "tty": true,
        "terminationMessagePolicy": "File"
    }]
}
```

To update the ephemeral containers of the already running `example-pod`:

```shell
kubectl replace --raw /api/v1/namespaces/default/pods/example-pod/ephemeralcontainers  -f ec.json

> ```

Reviewer 2: Is there a reason to show json vs yaml?

This will return the new list of ephemeral containers:

```json
{
   "kind":"EphemeralContainers",
   "apiVersion":"v1",
   "metadata":{
      "name":"example-pod",
      "namespace":"default",
      "selfLink":"/api/v1/namespaces/default/pods/example-pod/ephemeralcontainers",
      "uid":"a14a6d9b-62f2-4119-9d8e-e2ed6bc3a47c",
      "resourceVersion":"15886",
      "creationTimestamp":"2019-08-29T06:41:42Z"
   },
   "ephemeralContainers":[
      {
         "name":"debugger",
         "image":"busybox",
         "command":[
            "sh"
         ],
         "resources":{

         },
         "terminationMessagePolicy":"File",
         "imagePullPolicy":"IfNotPresent",
         "stdin":true,
         "tty":true
      }
   ]
}
```

The ephemeral container now appears in a `kubectl describe`:

> 

Reviewer 2: nit: reword line 157. The container appears in kubectl describe? Rather, the EC resource is created?

Reviewer 2: Okay now. The phrase `now appears in a` seemed odd to me.

```shell
kubectl describe pod example-pod
```

```
...
Ephemeral Containers:
  debugger:
    Container ID:  docker://cf81908f149e7e9213d3c3644eda55c72efaff67652a2685c1146f0ce151e80f
    Image:         busybox
    Image ID:      docker-pullable://busybox@sha256:9f1003c480699be56815db0f8146ad2e22efea85129b5b5983d0e0fb52d9ab70
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
    State:          Running
      Started:      Thu, 29 Aug 2019 06:42:21 +0000
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:         <none>
...
```

> And can be attached using `kubectl attach`:

Reviewer: ```suggestion
You can attach to the new ephemeral container using `kubectl attach`:
```

```shell
kubectl attach -it example-pod -c debugger
```

> When process namespace sharing is enabled, all pod processes are visible. For example:

Reviewer: ```suggestion
If process namespace sharing is enabled, you can see processes from all the containers in that Pod. For example:
```

```shell

> % kubectl attach -it example-pod -c debugger

Reviewer: nits:
1. Avoid including the prompt
2. Put the sample output in a separate block (and don't mark it as Bourne shell code)

Author: This is the separate block including the output (the command was in a previous block). I thought showing the two prompts (one outside the container and one inside), but I can remove the command from the output.

If you don't see a command prompt, try pressing enter.
/ # ps auxww
PID   USER     TIME  COMMAND
    1 root      0:00 /pause
    6 root      0:00 nginx: master process nginx -g daemon off;
   11 101       0:00 nginx: worker process
   12 101       0:00 nginx: worker process
   13 101       0:00 nginx: worker process
   14 101       0:00 nginx: worker process
   15 101       0:00 nginx: worker process
   16 101       0:00 nginx: worker process
   17 101       0:00 nginx: worker process
   18 101       0:00 nginx: worker process
   19 root      0:00 /pause
   24 root      0:00 sh
   29 root      0:00 ps auxww
```

{{% /capture %}}

{{% capture whatsnext %}}

* Learn how to easily [Debug a Running Pod](/docs/tasks/debug-application-cluster/debug-running-pod.md)

>   using `kubectl debug`.

Reviewer 2: nit: Learn how to easily [debug a running Pod](/docs/tasks/debug-application-cluster/debug-running-pod/) ...

### Examples

{{< note >}}
The examples in this section require the `EphemeralContainers` [feature

> gate](/docs/concepts/workloads/pods/ephemeral-containers/) to be

Reviewer 2: @verb, a few more links to fix.
Line 85, did you intend to link to the feature gates page, 
/docs/reference/command-line-tools-reference/feature-gates/

{{% capture whatsnext %}}

> * Learn how to easily [debug a running Pod](/docs/tasks/debug-application-cluster/debug-running-pod.md)

Reviewer 2: nit: please fix the link by removing the `.md`,
/docs/tasks/debug-application-cluster/debug-running-pod/
